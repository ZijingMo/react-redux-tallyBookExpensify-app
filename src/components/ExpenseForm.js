import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


//const now = moment();
//console.log(now.format("MMM Do, YYYY"));


// This state component creates an input form, date picker ,and a button for expense data record
export default class ExpenseForm extends React.Component {
  // Using the constructor function to access the expense data which has passed down.If not, the form just shows the default value
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calenderFocused: false,
      error: ''
    };
  }
  // These two functions below control the change of input value. It binds the input value and event handlers 'onChange'.
  // If user input some value, property 'description' or 'note' in object 'state' will be updated
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  // This function applies the reg-expression to stipulate the format of amount input, which is any number.two-digit number in this case
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createAt) => {
    if (createAt) {
      this.setState(() => ({ createAt }));
    }
  };
  // Destructuring the object from singleDatePicker component
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  }; 
  // Submitting data to Redux state container and error control
  onSubmit = (e) => {
    // Prevent refreshing the page when user clicks the 'Add Expense' button
    e.preventDefault(); 
    if (!this.state.description || !this.state.amount) {
      // Set error state 
      this.setState(() => ({ error: 'Please provide descriptions and amount.' }));
    } else {
      // Clear the error
      this.setState(() => ({ error: '' }));
      // Add the form (user's input data) to Redux store as an object
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createAt: this.state.createAt.valueOf(),// Using 'moment' library's build-in function valueOf() to convert the time in Unix Timestamp mode (milliseconds)
        note: this.state.note
      });
    }
  };
  render () {
    return (
      <div> 
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input 
            type = "text"
            placeholder ="Description"
            autoFocus
            value = {this.state.description}
            onChange = {this.onDescriptionChange}
          />
          <input 
            type = "text"
            placeholder = "Amount"
            value = {this.state.amount}
            onChange = {this.onAmountChange}
          />
          {/* Third party libraries' component to implement date-pick and calender select */}
          <SingleDatePicker 
            date={this.state.createAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            isDayHighlighted={() => true}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value = {this.state.note}
            onChange = {this.onNoteChange}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}