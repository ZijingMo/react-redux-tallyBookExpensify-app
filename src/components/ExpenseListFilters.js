import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

// This component implements value-input normally. It is able to update change of the input and fire the 'filter''s actions with the help of onChange() method
// The syntaxes of component 'DateRangePicker' follow the 'airbnb/react-dates' documentation
export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render () {
    return (
      <div>
        <input 
          type="text" 
          value={this.props.filters.text} 
          onChange={this.onTextChange} 
        />
    
        <select
          value={this.props.filters.sortBy} 
          onChange={this.onSortChange}
        >
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>

        <DateRangePicker 
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
    filters: state.filters 
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);