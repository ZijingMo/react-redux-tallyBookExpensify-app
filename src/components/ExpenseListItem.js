import React from 'react';
import { Link } from 'react-router-dom';

// Destructuring the object to 'dispatch', 'id', 'description', 'amount', 'createAt'
// The component is injected to the 'dispatch' function when we use connect() hoc to wrap it up
const ExpenseListItem = ({ id, description, amount, createAt }) => (
  <div>
    {/* Import 'Link' component to redirect edit page */}
    <Link to={`/edit/${id}`}>
       <h2> { description } </h2> 
    </Link>
    <p> {amount} - {createAt} </p>
  </div>
);

export default ExpenseListItem; 