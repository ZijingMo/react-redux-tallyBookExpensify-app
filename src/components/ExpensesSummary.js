import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ( { expenseCount, expensesTotal } ) => {
  // Format change about word 'expense' and 'expenses'
  const SingularAndPluralChange = expenseCount === 1 || !expenseCount ? 'expense' : 'expenses';
  // Format change from number to money counting
  const formattedExpense = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Hmm.. You have <span>{expenseCount}</span> {SingularAndPluralChange}, and the total is <span>{formattedExpense}</span> </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const VisibleExpenses = getVisibleExpenses(state.expenses, state.filters);

  return {
    // After filtering from method 'getVisibleExpenses()', the amount of objects in this array could be regarded as the count
    expenseCount: VisibleExpenses.length,
    expensesTotal: getExpensesTotal(VisibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);