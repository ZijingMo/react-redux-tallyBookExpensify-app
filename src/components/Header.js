import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth'; 

// Destruct the object 'startLogout', which is a function  
export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__contentStyle">
        <Link className="header__title" to="/dashboard">
          <h1>Expensify - A Tally Book</h1>
        </Link>
        <button className="button--link" onClick = {startLogout}> Logout </button>
      </div>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

// It is not necessary to connect to the state, so the first argument is undefined 
export default connect(undefined, mapDispatchToProps)(Header);