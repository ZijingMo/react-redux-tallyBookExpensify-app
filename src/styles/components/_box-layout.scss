.box-layout {
  align-items: center;
  background: url('/images/moneyBG.gif'); //piggyBank.jpg, moneyBG.gif
  background-size: 100% 100%; // 50% 100%
  background: cover;
  display: flex;
  height: 100vh;
  justify-content: center;
  width: 100vw;
  //position: absolute;
}

.box-layout__box {
  //display: block;
  background: fade-out(aliceblue, 0.1);
  border-radius: 10px;
  padding: 3.2rem, 1.6rem, 3.2rem, 1.6rem;
  text-align: center;
  width: 50rem;
}

.box-layout__title {
  margin: 0 0 $m-size, 0;
  line-height: 1;
  font-size: 4rem;
  
}

.box-layout__message {
  margin: 0 $m-size 0 $m-size;
  font-size: 2rem;
}