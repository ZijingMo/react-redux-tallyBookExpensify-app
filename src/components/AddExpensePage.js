import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// Refactoring the code, from stateless to state component for jest test
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    // This method leads to redirect to the expensify dashboard (Home-Page) when the user click the button 'Add Expense' 
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container"> 
          <ExpenseForm 
            onSubmit = {this.onSubmit}
          />        
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);