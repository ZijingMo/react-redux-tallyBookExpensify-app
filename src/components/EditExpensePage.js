import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

// Refactoring the code, from functional to class component:
// For jest test

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    // This method leads to redirect to the expensify dashboard (Home-Page) when the user click the button 'Add Expense' 
    this.props.history.push('/');
    //console.log('updated', expense);
  };

  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense = {this.props.expense} 
          onSubmit = {this.onSubmit}
        />
        <button 
          onClick = {this.onRemove}
        > 
            Remove 
        </button>
    </div>
    );
  }
}

// New component will subscribe to Redux store updates. 
// This means that any time the store is updated, mapStateToProps will be called. 
// The results of mapStateToProps must be a plain object, which will be merged into the component’s props.
const mapStateToProps = (state, props) => {
  return {
    // It is going to search the array 'expenses' (a lot of objects 'expense' form array 'expenses') in Redux store to match the updated expense the user input
    // 'expense.id' means the id of array 'expenses' in Redux store
    // 'prop.match.params.id' means the property 'id' of an updated object 
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

// Every action creator wrapped into a dispatch call so they may be invoked directly, will be merged into the component’s props.
// Import actions (editExpense, removeExpense) from action folder and subscribe to prop in class 'EditExpensePage'
const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data)) 
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);



// Functional form as a reference

// const EditExpensePage = (props) => {
//   //console.log(props);
//   return (
//     <div>
//       <ExpenseForm
//         expense={props.expense} 
//         onSubmit={(expense) => {
//           // Dispatch the action to edit the expense and redirect to the dashboard
//           props.dispatch(editExpense(props.expense.id, expense));
//           props.history.push('/');
//           console.log('updated', expense);
//         }}
//       />
//       <button 
//         onClick={() => {
//           props.dispatch(removeExpense({id: props.expense.id})); //The argument of action function removeExpense() is an object
//           props.history.push('/');
//         }}
//       > 
//           Remove 
//       </button>
//     </div>
//   );
// };

