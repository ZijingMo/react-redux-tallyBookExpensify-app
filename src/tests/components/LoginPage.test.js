import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage';

test('should render component LoginPage correctly', () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startLogin button on button click correctly', () => {
  const startLogin = jest.fn() // Using mock functions
  const wrapper = shallow(<LoginPage startLogin={startLogin} />);
  wrapper.find('button').simulate('click');
  expect(startLogin).toHaveBeenCalled();
});

