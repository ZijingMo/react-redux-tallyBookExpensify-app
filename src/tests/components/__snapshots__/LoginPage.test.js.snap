// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component LoginPage correctly 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Expensify
    </h1>
    <p
      className="box-layout__message"
    >
      "Money is a terrible master but an excellent servant." â€“ P.T. Barnum.
    </p>
    <button
      className="button"
    >
       Login with Google
    </button>
  </div>
</div>
`;
